File: src/main/java/infinitespire/relics/crystals/FocusingShard.java
package infinitespire.relics.crystals;

import com.megacrit.cardcrawl.actions.GameActionManager;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.cards.CardGroup;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.vfx.cardManip.ShowCardBrieflyEffect;
import infinitespire.InfiniteSpire;
import infinitespire.abstracts.InfiniteRelic;

public class FocusingShard extends InfiniteRelic{
    public static final String ID = InfiniteSpire.createID("FocusingShard");

    public FocusingShard(){
        super(ID, "focusingshard", LandingSound.CLINK);
    }

    @Override
    public void atBattleStartPreDraw() {
        GameActionManager manager = AbstractDungeon.actionManager;
        AbstractPlayer player = AbstractDungeon.player;
        CardGroup drawPile = player.drawPile.getUpgradableCards();

        for(int i = 0; i < this.counter; i++) {
            AbstractCard card = drawPile.group.get(AbstractDungeon.miscRng.random(drawPile.size() - 1));
            card.upgrade();
            AbstractDungeon.effectList.add(new ShowCardBrieflyEffect(card.makeStatEquivalentCopy()));
        }
    }
}

File: src/main/kotlin/com/blanktheevil/infinitespire/relics/FocusingShard.kt
package com.blanktheevil.infinitespire.relics

import com.blanktheevil.infinitespire.extensions.getRandomItem
import com.blanktheevil.infinitespire.extensions.makeID
import com.blanktheevil.infinitespire.extensions.player
import com.blanktheevil.infinitespire.relics.abstracts.CrystalRelic
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.dungeons.AbstractDungeon.miscRng
import com.megacrit.cardcrawl.vfx.cardManip.ShowCardBrieflyEffect

class FocusingShard : CrystalRelic(ID, IMG) {
  companion object {
    val ID = "FocusingShard".makeID()
    private const val IMG = "focusingshard"
  }

  // fairly certain this will crash or not work as expected
  override fun atBattleStartPreDraw() {
    player.drawPile.upgradableCards.group.forEachIndexed { index: Int, _: AbstractCard ->
      if (counter > index) {
        player.drawPile.upgradableCards.group.getRandomItem(miscRng)?.also {
          it.upgrade()
          AbstractDungeon.effectList.add(
            ShowCardBrieflyEffect(it.makeStatEquivalentCopy())
          )
        }
      }
    }
  }
}